### PR TITLE
feat: use dontBatch directive in algolia

### PR DESCRIPTION
**Pull request by aashishmalik**
## What are the changes introduced in this PR?

Onboard Algolia destination to transformer proxy and update batching logic to use dontBatch in retry.

## What is the related Linear task?

Resolves INT-1697

## Please explain the objectives of your changes below

1. Converting all the events's metadata and adding `dontBatch` to true.
2. In the next iteration, these events are sent in isloation.
3.  If again on retry for the event sent in isolation, recieved code `422`, aborting these events to stop from retrying continously.

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3169/test-report.html

File: src/cdk/v2/destinations/algolia/rtWorkflow.yaml [hunk 1]
```diff
@@ -28,10 +27,47 @@ steps:
       $.outputs.transform#idx.error.(
         $.handleRtTfSingleEventError(^[idx], .originalError ?? ., {})
       )[]
-  - name: batchSuccessfulEvents
+
+  - name: separateDontBatchEventForProxy
```
Comment (koladilip) [inline src/cdk/v2/destinations/algolia/rtWorkflow.yaml:31]:
```
bindings:
  - path: ../../../../v0/destinations/algolia/config
  - name: handleRtTfSingleEventError
    path: ../../../../v0/util/index
steps:
  - name: validateInput
    template: |
      $.assert(Array.isArray(^) && ^.length > 0, "Invalid event array")

  - name: transform
    externalWorkflow:
      path: ./procWorkflow.yaml
      bindings:
        - name: batchMode
          value: true
    loopOverInput: true

  - name: successfulEvents
    template: |
      $.outputs.transform#idx.output.({
        "output": .,
        "destination": ^[idx].destination,
        "metadata": ^[idx].metadata
      })[]
  - name: failedEvents
    template: |
      $.outputs.transform#idx.error.(
        $.handleRtTfSingleEventError(^[idx], .originalError ?? ., {})
      )[]

  - name: batchSuccessfulEvents
    description: Batches the successfulEvents
    template: |
      const dontBatchTrueEvents =  $.outputs.successfulEvents{.metadata.dontBatch}[];
      const dontBatchFalseEvents = $.outputs.successfulEvents{!.metadata.dontBatch}[];
      
      let batches = [...$.chunk(dontBatchFalseEvents, $.MAX_BATCH_SIZE), ...dontBatchTrueEvents];
      batches@batch.({
        "batchedRequest": {
          "body": {
            "JSON": {"events": ~r batch.output[]},
            "JSON_ARRAY": {},
            "XML": {},
            "FORM": {}
          },
          "version": "1",
          "type": "REST",
          "method": "POST",
          "endpoint": {{$.ENDPOINT}},
          "headers": batch[0].destination.Config.().({
            "X-Algolia-Application-Id": .applicationId,
            "X-Algolia-API-Key": .apiKey
          }),
          "params": {},
          "files": {}
        },
        "metadata": ~r batch.metadata[],
        "batched": true,
        "statusCode": 200,
        "destination": batch[0].destination
      })[];

  - name: finalPayload
    template: |
      [...$.outputs.failedEvents, ...$.outputs.batchSuccessfulEvents]
```

try this, there are few tests failing but you will get idea

Comment (aashishmalik) [reply src/cdk/v2/destinations/algolia/rtWorkflow.yaml]: done

**Review by koladilip — APPROVED** [review]

File: src/cdk/v2/destinations/algolia/rtWorkflow.yaml [hunk 2]
```diff
@@ -28,10 +27,14 @@ steps:
       $.outputs.transform#idx.error.(
         $.handleRtTfSingleEventError(^[idx], .originalError ?? ., {})
       )[]
+
   - name: batchSuccessfulEvents
     description: Batches the successfulEvents
     template: |
-      let batches = $.chunk($.outputs.successfulEvents, $.MAX_BATCH_SIZE);
+      const dontBatchTrueEvents =  $.outputs.successfulEvents{.metadata.dontBatch}[];
+      const dontBatchFalseEvents = $.outputs.successfulEvents{!.metadata.dontBatch}[];
```
Comment (ItsSudip) [inline src/cdk/v2/destinations/algolia/rtWorkflow.yaml:35]:
```suggestion
      const dontBatchEvents =  $.outputs.successfulEvents{.metadata.dontBatch}[];
      const batchEvents = $.outputs.successfulEvents{!.metadata.dontBatch}[];
```

File: src/v1/destinations/algolia/networkHandler.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    return {
+      status,
+      message,
+      destinationResponse,
+      response: responseWithIndividualEvents,
+    };
+  }
+
+  // in case of non 2xx status sending 500 for every event, populate response and update dontBatch to true
```
Comment (ItsSudip) [inline src/v1/destinations/algolia/networkHandler.js:42]: This statement is not correct we are returning 400 when `dontBatch` is true and status is 422.

**Review by mihir-4116 — APPROVED** [review]

**Review by ItsSudip — COMMENTED** [review]
We need to add test cases for proxy.

File: src/v1/destinations/algolia/networkHandler.js (hunk 3 above)
Comment (aashishmalik) [reply src/v1/destinations/algolia/networkHandler.js:42]: we sending 400 as 422 won't be handled in server, 500 will be sent first time

File: src/cdk/v2/destinations/algolia/rtWorkflow.yaml (hunk 2 above)
Comment (aashishmalik) [reply src/cdk/v2/destinations/algolia/rtWorkflow.yaml:35]: going with this implementation to distinguish with flag naming

File: src/v1/destinations/algolia/networkHandler.js (hunk 3 above)
Comment (ItsSudip) [reply src/v1/destinations/algolia/networkHandler.js:42]: Can you update the statement then accordingly?

**Pull request merged**